pkg/raft/raft.go

```diff
@@ -1702,7 +1706,22 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 	})
 }
 
+// checkMatch ensures that the follower's log size does not contradict to the
+// leader's idea where it matches.
+func (r *raft) checkMatch(match uint64) {
+	// TODO(pav-kv): lastIndex() might be not yet durable. Make this check
+	// stronger by comparing `match` with the last durable index.
+	//
+	// TODO(pav-kv): make this check stronger when the raftLog stores the last
+	// accepted term. If `match` is non-zero, this follower's log last accepted
+	// term must equal the leader term, and have entries up to `match` durable.
+	if last := r.raftLog.lastIndex(); last < match {
+		r.logger.Panicf("match(%d) is out of range [lastIndex(%d)]. Was the raft log corrupted, truncated, or lost?", match, last)
```

Comment:
Should we add a test for the MsgApp and MsgHeartbeat cases?

Reply:
Done.